### PR TITLE
Fixes for the JsonFraming stage

### LINKED ISSUE #21150: JsonFraming.objectScanner failed to emit all output objects for some input - 2.4.9-RC2
It seems that for certain ByteString inputs, the `JsonFraming.objectScanner` fails to reliably emit all objects. I noticed this when trying to add tests for #21147. For example, the following spec reliably fails for me:

```scala
    "parse an array of small objects" in {
      val input = """[{"a": 0}, {"b":1}, {"c":2}, {"d":3}, {"e":4}]"""

      val result = Source.single(ByteString(input))
        .via(JsonFraming.objectScanner(Int.MaxValue))
        .runFold(0) {
          case (acc, _) ⇒ acc + 1
        }
      result.futureValue shouldBe 5
    }
```

The output count is sometimes 3, sometimes 4 objects, but that last elements do not seem to be consumed.

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #21150, #21154 and #21150

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/javadsl/JsonFraming.scala [hunk 1]
```diff
@@ -29,8 +29,6 @@ object JsonFraming {
    * elements are separated by multiple newlines or other whitespace characters. And of course is insensitive
    * (and does not impact the emitting frame) to the JSON object's internal formatting.
    *
-   * Framing raw JSON values (such as integers or strings) is supported as well.
-   *
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/JsonFraming.scala]: #21151

File: akka-stream/src/main/scala/akka/stream/scaladsl/JsonFraming.scala [hunk 2]
```diff
@@ -36,16 +36,17 @@ object JsonFraming {
...
    *                            this Flow will fail the stream.
    */
   def objectScanner(maximumObjectLength: Int): Flow[ByteString, ByteString, NotUsed] =
     Flow[ByteString].via(new SimpleLinearGraphStage[ByteString] {
-      private[this] val buffer = new JsonObjectParser(maximumObjectLength)
+
+      override protected def initialAttributes: Attributes = Attributes.name("JsonFraming.objectScanner")
 
       override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with InHandler with OutHandler {
+        private val buffer = new JsonObjectParser(maximumObjectLength)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/JsonFraming.scala]: #21154 buffer was shared

File: akka-stream/src/main/scala/akka/stream/scaladsl/JsonFraming.scala [hunk 3]
```diff
@@ -57,8 +58,8 @@ object JsonFraming {
           tryPopBuffer()
 
         override def onUpstreamFinish(): Unit = {
-          try buffer.poll() match {
-            case Some(json) ⇒ emit(out, json, () ⇒ completeStage())
+          buffer.poll() match {
+            case Some(json) ⇒ emit(out, json)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/JsonFraming.scala]: #21150 stage completed after emitting one element when upstream closes (even if there was more elements available in the buffer)

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/JsonFramingSpec.scala [hunk 4]
```diff
@@ -116,6 +117,34 @@ class JsonFramingSpec extends AkkaSpec {
...
+      output.expectNext("""{"b":1}""")
+      output.request(1)
+      output.expectNext("""{"c":2}""")
+      output.request(1)
+      output.expectNext("""{"d":3}""")
+      output.request(1)
+      output.expectNext("""{"e":4}""")
+      output.request(1)
+      output.expectComplete()
+    }
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/JsonFramingSpec.scala]: Test coverage for #21150

File: akka-stream/src/main/scala/akka/stream/scaladsl/JsonFraming.scala [hunk 5]
```diff
@@ -36,16 +36,17 @@ object JsonFraming {
...
-   * Framing raw JSON values (such as integers or strings) is supported as well.
-   *
    * @param maximumObjectLength The maximum length of allowed frames while decoding. If the maximum length is exceeded
    *                            this Flow will fail the stream.
    */
   def objectScanner(maximumObjectLength: Int): Flow[ByteString, ByteString, NotUsed] =
     Flow[ByteString].via(new SimpleLinearGraphStage[ByteString] {
-      private[this] val buffer = new JsonObjectParser(maximumObjectLength)
+
+      override protected def initialAttributes: Attributes = Attributes.name("JsonFraming.objectScanner")
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/JsonFraming.scala]: boyscouting

**Comment by drewhk** [comment]
LGTM

**Comment by patriknw** [comment]
LGTM (3 flugor in en smäll)

**Pull request merged**